cns/kubecontroller/podwatcher/podwatcher.go

```diff
@@ -0,0 +1,66 @@
+package podwatcher
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type podcli interface {
+	List(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error
+}
+
+type podListener interface {
+	Update([]v1.Pod)
+}
+
+type podWatcher struct {
+	cli       podcli
+	listOpt   client.ListOption
+	listeners []podListener
+}
+
+func New(nodename string, listeners ...podListener) *podWatcher { //nolint:revive // private struct to force constructor
+	return &podWatcher{
+		listeners: listeners,
+		listOpt:   &client.ListOptions{FieldSelector: fields.SelectorFromSet(fields.Set{"spec.nodeName": nodename})},
+	}
+}
+
+func (p *podWatcher) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	podList := &v1.PodList{}
+	if err := p.cli.List(ctx, podList, p.listOpt); err != nil {
```

Comment:
is calling List for each pod event necessary? could we just keep track of each pod individually, and if CNS doesn't know about a pod when a request comes in, we just fail the request?

Reply:
I wrote this keeping in mind that the podwatcher could be reused by regular swift scenario in the ipam redesign...
We don't need to List in v2, wiring up the reconciler is sufficient, so I can make this a little more middleware-ish and dynamically inject the whole reconcile func maybe?

Reply:
did some abstracting here - I went ahead and wrote some boilerplate for future use, but the net current behavior is that reconcile events are discarded for V2 and we are only using this to populate the cache.